src/librustc/ty/layout.rs

```diff
@@ -1490,27 +1492,34 @@ impl<'gcx, 'tcx, T: HasTyCtxt<'gcx>> HasTyCtxt<'gcx> for LayoutCx<'tcx, T> {
     }
 }
 
-pub trait MaybeResult<T> {
+pub trait MaybeResult<T, E> {
     fn from_ok(x: T) -> Self;
     fn map_same<F: FnOnce(T) -> T>(self, f: F) -> Self;
+    fn to_result(self) -> Result<T, E>;
 }
 
-impl<T> MaybeResult<T> for T {
+impl<T, E> MaybeResult<T, E> for T {
```

Comment:
Instead of a generic parameter you can use `!` for the associated type here, since we know that `to_result` can only result in an `Ok`

Reply:
I tried that earlier. I get this:
```
error: expected one of `>`, `const`, identifier, or lifetime, found `!`118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1501:9
     |
1501 | impl<T, !> MaybeResult<T, !> for T {
     |         ^ expected one of `>`, `const`, identifier, or lifetime here

error: aborting due to previous error

```

Reply:
```
error[E0107]: wrong number of type arguments: expected 1, found 2    ] 118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1501:24
     |
1501 | impl<T> MaybeResult<T, !> for T {
     |                        ^ unexpected type argument

```

Reply:
I'll read a bit and try to figure these out.

Reply:
you can just leave both of these out. `!` is a concrete type. Use it for the associated type directly without giving a second generic parameter